group/assignment5/main.cpp

```diff
@@ -0,0 +1,194 @@
+        if (word.size() == 0) {
+            continue;
+        }
+        char currentChar = word[0];
+        if (currentChar != prevChar) {
+            if (suffixesByCharacters.find(currentChar) != suffixesByCharacters.end()) { // found a circle in characters - inconsistent dictionary
+                return false;
+            }
+            // mark prevChar as prior to currentChar and add new suffixes vector for current char
+            charPriorityGraph[prevChar][currentChar] = true;
+            suffixesByCharacters[currentChar] = vector<string>(0);
+            numOfPriorCharacters[currentChar]++;
+        }
+        word.erase(0, 1);
+        suffixesByCharacters[currentChar].push_back(word);
+        prevChar = currentChar;
+    }
+    for (auto it : suffixesByCharacters) {
+        vector<string> suffixes = it.second;
+        createCharPriorityGraph(suffixes, charPriorityGraph, numOfPriorCharacters);
+    }
+    return true;
+}
+
+bool findNextOrigin(const unordered_map<char, int> &numOfPriorCharacters, char &origin) {
+    for (auto it : numOfPriorCharacters) {
+        if (it.second == 0) {
+            origin = it.first;
+            return true;
+        }
+    }
+    // no origin -> circle in graph -> dictionary is inconsistent
+    return false;
+}
+
+void updateNumOfPriorCharacters(const unordered_map<char, vector<bool>> &charPriorityGraph, unordered_map<char, int> &numOfPriorCharacters, const char origin) {
+    const vector<bool> children = charPriorityGraph.at(origin);
+    for (char c = 0; c < children.size(); c++) {
+        if (children[c]) {
+            numOfPriorCharacters[c]--;
+        }
+    }
+    numOfPriorCharacters.erase(origin);
+}
+
+bool findAlphabet(vector<string> &dictionary, vector<char> &alphabet) {
+    vector<char> alphabetCharacters = findAlphabetCharacters(dictionary);
+
+    unordered_map<char, vector<bool>> charPriorityGraph;
+    unordered_map<char, int> numOfPriorCharacters;
+    for (auto c : alphabetCharacters) {
+        charPriorityGraph[c] = vector<bool>((unsigned int)pow(2, CHAR_BIT), false);
+        numOfPriorCharacters[c] = 0;
+    }
+
+    if (!createCharPriorityGraph(dictionary, charPriorityGraph, numOfPriorCharacters)) {
+        alphabet.clear();
+        return false;
+    }
+
+    // find topological sort of characters graph to find alphabet
+    alphabet.clear();
+    char nextChar;
+    while(numOfPriorCharacters.size() > 0) {
+        if (!findNextOrigin(numOfPriorCharacters, nextChar)) {
+            alphabet.clear();
+            return false;
+        }
+        alphabet.push_back(nextChar);
+        updateNumOfPriorCharacters(charPriorityGraph, numOfPriorCharacters, nextChar);
+    }
+    return true;
+}
+
+int main() {
+    vector<string> dictionary = {"ART", "RAT", "CAT", "CAR"};
+    vector<char> alphabet;
+    bool result;
+    unordered_map<char, int> idxs;
+    vector<bool> expectedChars;
+    vector<bool> actualChars;
+
+    // test 1 - assignment example
```

Comment:
Point for a discussion here: I really love the way you did the testing, because there are a lot of cases in this problem in which the right answer is not unique. So you do a general validation which can be applied even if the algorithm will change. However if we are talking about this particular implementation, it is known what exactly alphabet from a set of potential right ones we will get. Maybe it should be checked that our expectations are met?